src/components/Tabs/ControlledTabs.js

```diff
@@ -0,0 +1,211 @@
+     */
+    onChangeTab: PropTypes.func.isRequired,
+    /**
+     * When this is true, inActive tabs will be rendered and hidden with .slds-hide instead of returning null
+     */
+    renderInactiveTabs: PropTypes.bool,
+    /**
+     * Renders scoped variant
+     */
+    scoped: PropTypes.bool,
+    /**
+     * Renders Card-like variant
+     */
+    styled: PropTypes.bool,
+  };
+
+  static getPanelState(children) {
+    const positionReducer = (acc, curr, i) => ({
+      ...acc,
+      [`${curr.props.id}`]: i,
+    });
+
+    const childrenArray = React.Children.toArray(children);
+
+    const nextState = {
+      positions: childrenArray.reduce(positionReducer, {}),
+      focusedTab: false,
+      length: childrenArray.length,
+    };
+
+    return nextState;
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = ControlledTabs.getPanelState(props.children);
+    /** refs are used to focus links after keyboard navigation occured */
+    this.linkNodes = new Map();
+  }
+
+  componentWillReceiveProps({ children: nextChildren }) {
+    const { children: prevChildren } = this.props;
+
+    if (nextChildren !== prevChildren) {
+      this.setState(ControlledTabs.getPanelState(nextChildren));
+    }
+  }
+
+  onChangeTab(nextActiveTab) {
+    const { activeTab, onChangeTab } = this.props;
+    if (nextActiveTab !== activeTab) {
+      this.setState({ focusedTab: nextActiveTab });
+      onChangeTab(nextActiveTab);
+    }
+  }
+
+
+  getCurrentPosition() {
+    const { activeTab } = this.props;
+    const { positions } = this.state;
+    return positions[activeTab];
+  }
+
+  getIdForPosition(position) {
+    const { positions } = this.state;
+    return Object.keys(positions).find(key => positions[key] === position);
+  }
+
+  getOnLinkKeyup() {
+    const linkNodes = this.linkNodes;
+
+    return (evt) => {
+      const { key } = evt;
+      const isLeftKey = key === 'ArrowLeft';
+      const isRightKey = key === 'ArrowRight';
+      if (!isLeftKey && !isRightKey) return;
+
+      const { length } = this.state;
+      const currentPosition = this.getCurrentPosition();
+
+      /* eslint-disable no-nested-ternary */
+      const nextIndex = isLeftKey
+        ? (currentPosition > 0) ? currentPosition - 1 : Math.max(length - 1, 0)
+        : (currentPosition < length - 1) ? currentPosition + 1 : 0;
+      /* eslint-enable */
+
+      const nextId = this.getIdForPosition(nextIndex);
+
+      this.onChangeTab(nextId);
+      /** Accesses the ref defined in <TabLink /> */
+      const linkNode = linkNodes.get(nextId);
+      if (linkNode && linkNode.link) { linkNode.link.focus(); }
+    };
+  }
+
+  getLinkRefSetter(id) {
+    return (c) => { this.linkNodes.set(id, c); };
+  }
+
+  getOnLinkFocus(id) {
+    return () => { this.setState({ focusedTab: id }); };
+  }
+
+  getOnBlur() {
+    return ({ currentTarget }) => {
+      /** Workaround to only fire blur when the whole navigation is unfocused */
+      setTimeout(() => {
```

Comment:
This workaround doesn't seem to make a big difference to me (there is an extra setState() but it doesn't seem to cause a re-render, probably merged with the `setState()` in `getOnLinkFocus` afterwards or so ...).

Reply:
Firefox? It's working in Safari & Chrome for me, Firefox has a 10 yo bug relating to this :/

Reply:
Yes, i was testing in Firefox. Removing the setTimeout() didn't make a difference there so i was wondering whether i was missing some edge-case ...